File: apps/mobile/lib/models/agent.dart
class PermissionPolicy {
  final List<String> allowedHosts;
  final List<String> allowedNetworks;
  final bool allowFileAccess;
  final bool allowLoopback;
  final int maxNodes;
  final int maxLoops;
  final int maxRequestSize;
  final int maxResponseSize;
  final int maxExecutionSeconds;

  const PermissionPolicy({
    this.allowedHosts = const [],
    this.allowedNetworks = const [],
    this.allowFileAccess = false,
    this.allowLoopback = false,
    this.maxNodes = 50,
    this.maxLoops = 10,
    this.maxRequestSize = 10485760,
    this.maxResponseSize = 10485760,
    this.maxExecutionSeconds = 600,
  });

  factory PermissionPolicy.fromJson(Map<String, dynamic> json) {
    return PermissionPolicy(
      allowedHosts: List<String>.from(json['allowed_hosts'] ?? []),
      allowedNetworks: List<String>.from(json['allowed_networks'] ?? []),
      allowFileAccess: json['allow_file_access'] ?? false,
      allowLoopback: json['allow_loopback'] ?? false,
      maxNodes: json['max_nodes'] ?? 50,
      maxLoops: json['max_loops'] ?? 10,
      maxRequestSize: json['max_request_size'] ?? 10485760,
      maxResponseSize: json['max_response_size'] ?? 10485760,
      maxExecutionSeconds: json['max_execution_seconds'] ?? 600,
    );
  }

  Map<String, dynamic> toJson() => {
    'allowed_hosts': allowedHosts,
    'allowed_networks': allowedNetworks,
    'allow_file_access': allowFileAccess,
    'allow_loopback': allowLoopback,
    'max_nodes': maxNodes,
    'max_loops': maxLoops,
    'max_request_size': maxRequestSize,
    'max_response_size': maxResponseSize,
    'max_execution_seconds': maxExecutionSeconds,
  };
}

class Agent {
  final String id;
  final String name;
  final String? description;
  final String? systemPrompt;
  final String? model;
  final double? temperature;
  final int? maxTokens;
  final PermissionPolicy permissions;
  final DateTime createdAt;
  final DateTime updatedAt;

  const Agent({
    required this.id,
    required this.name,
    this.description,
    this.systemPrompt,
    this.model,
    this.temperature,
    this.maxTokens,
    required this.permissions,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Agent.fromJson(Map<String, dynamic> json) {
    return Agent(
      id: json['id'] as String,
      name: json['name'] as String,
      description: json['description'] as String?,
      systemPrompt: json['system_prompt'] as String?,
      model: json['model'] as String?,
      temperature: (json['temperature'] as num?)?.toDouble(),
      maxTokens: json['max_tokens'] as int?,
      permissions: PermissionPolicy.fromJson(json['permissions'] ?? {}),
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
    );
  }

  Map<String, dynamic> toJson() => {
    'id': id,
    'name': name,
    'description': description,
    'system_prompt': systemPrompt,
    'model': model,
    'temperature': temperature,
    'max_tokens': maxTokens,
    'permissions': permissions.toJson(),
    'created_at': createdAt.toIso8601String(),
    'updated_at': updatedAt.toIso8601String(),
  };
}

class AgentInput {
  final String name;
  final String? description;
  final String? systemPrompt;
  final String? model;
  final double? temperature;
  final int? maxTokens;
  final PermissionPolicy? permissions;

  const AgentInput({
    required this.name,
    this.description,
    this.systemPrompt,
    this.model,
    this.temperature,
    this.maxTokens,
    this.permissions,
  });

  Map<String, dynamic> toJson() => {
    'name': name,
    'description': description,
    'system_prompt': systemPrompt,
    'model': model,
    'temperature': temperature,
    'max_tokens': maxTokens,
    'permissions': permissions?.toJson(),
  };
}

File: crates/core/src/models/agent.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

/// Permission policy that controls what an agent/workflow is allowed to do.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PermissionPolicy {
    /// Allowed HTTP hostnames (wildcard * supported).
    #[serde(default)]
    pub allowed_hosts: Vec<String>,
    /// Allowed network CIDR ranges.
    #[serde(default)]
    pub allowed_networks: Vec<String>,
    /// Whether file:// access is allowed.
    #[serde(default)]
    pub allow_file_access: bool,
    /// Whether loopback/localhost access is allowed.
    #[serde(default)]
    pub allow_loopback: bool,
    /// Maximum number of nodes in a workflow.
    #[serde(default = "default_max_nodes")]
    pub max_nodes: i32,
    /// Maximum number of loop iterations.
    #[serde(default = "default_max_loops")]
    pub max_loops: i32,
    /// Maximum HTTP request body size in bytes.
    #[serde(default = "default_max_size")]
    pub max_request_size: i64,
    /// Maximum HTTP response body size in bytes.
    #[serde(default = "default_max_size")]
    pub max_response_size: i64,
    /// Maximum workflow execution time in seconds.
    #[serde(default = "default_max_execution_seconds")]
    pub max_execution_seconds: i64,
}

const fn default_max_nodes() -> i32 {
    50
}
const fn default_max_loops() -> i32 {
    10
}
const fn default_max_size() -> i64 {
    10 * 1024 * 1024
} // 10 MB
const fn default_max_execution_seconds() -> i64 {
    600
} // 10 minutes

impl Default for PermissionPolicy {
    fn default() -> Self {
        Self {
            allowed_hosts: Vec::new(),
            allowed_networks: Vec::new(),
            allow_file_access: false,
            allow_loopback: false,
            max_nodes: default_max_nodes(),
            max_loops: default_max_loops(),
            max_request_size: default_max_size(),
            max_response_size: default_max_size(),
            max_execution_seconds: default_max_execution_seconds(),
        }
    }
}

/// An AI Agent configuration.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Agent {
    pub id: Uuid,
    pub name: String,
    pub description: Option<String>,
    pub system_prompt: Option<String>,
    pub model: Option<String>,
    pub temperature: Option<f64>,
    pub max_tokens: Option<i32>,
    pub permissions: PermissionPolicy,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

/// Input for creating or updating an Agent.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentInput {
    pub name: String,
    pub description: Option<String>,
    pub system_prompt: Option<String>,
    pub model: Option<String>,
    pub temperature: Option<f64>,
    pub max_tokens: Option<i32>,
    pub permissions: Option<PermissionPolicy>,
}
